src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1215,59 +1215,84 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
+                    break;
+                case BoundNullCoalescingOperatorResultKind.LeftUnwrappedRightType:
+                    resultType = TypeSymbolWithAnnotations.Create(getRightResultType(unwrapIfNullable(leftResult.Type.TypeSymbol), rightResult.Type.TypeSymbol));
+                    break;
+                case BoundNullCoalescingOperatorResultKind.RightDynamicType:
+                    resultType = TypeSymbolWithAnnotations.Create(rightResult.Type.TypeSymbol, resultIsNullable);
+                    break;
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(node.OperatorResultKind);
             }
-            else
-            {
-                // PROTOTYPE(NullableReferenceTypes): Conversions: NullCoalescingOperator
-#if DEBUG
-                //Debug.Assert((object)leftType == null || AreCloseEnough(leftType.TypeSymbol, node.Type));
-                //Debug.Assert((object)rightType == null || AreCloseEnough(rightType.TypeSymbol, node.Type));
-#endif
 
-                // PROTOTYPE(NullableReferenceTypes): Capture in BindNullCoalescingOperator
-                // which side provides type and use that to determine nullability.
-                resultType = TypeSymbolWithAnnotations.Create((leftType ?? rightType)?.TypeSymbol, isNullableIfReferenceType: rightType?.IsNullable & leftType?.IsNullable);
+            _result = resultType;
+            return null;
 
-                ReportNullabilityMismatchIfAny(node.LeftOperand, resultType, leftType);
-                ReportNullabilityMismatchIfAny(node.RightOperand, resultType, rightType);
+            bool? getIsNullable(BoundExpression e, Result r) => (r.Type is null) ? e.IsNullable() : r.Type.IsNullable;
+            TypeSymbol getLeftResultType(TypeSymbol leftType, TypeSymbol rightType)
+            {
+                // If there was an identity conversion between the two operands (in short, if
+                // there is no conversion on the right operand), then check nullable conversions
+                // in both directions since it's possible the right operand is the better result type.
+                if ((object)rightType != null &&
+                    node.RightOperand.Kind != BoundKind.Conversion &&
+                    GenerateConversionForConditionalOperator(node.LeftOperand, leftType, rightType, reportMismatch: false))
+                {
+                    return rightType;
+                }
+                GenerateConversionForConditionalOperator(node.RightOperand, rightType, leftType, reportMismatch: true);
+                return leftType;
+            }
+            TypeSymbol getRightResultType(TypeSymbol leftType, TypeSymbol rightType)
+            {
+                GenerateConversionForConditionalOperator(node.LeftOperand, leftType, rightType, reportMismatch: true);
+                return rightType;
+            }
+            TypeSymbol unwrapIfNullable(TypeSymbol type)
```

Comment:
>unwrapIfNullable [](start = 23, length = 16)

I think there is an existing helper to do that, called "StrippedType" or something similar.